### PR TITLE
vulkan: fix SIGSEGV on config window opening

### PR DESCRIPTION
**Pull request by isilence**
Racy stores to vk{FunctionName} pointers makes dolphin crush on opening graphics config window during rendering. The fix -- make private copy of this pointers when necessary. 

Alternatively, we can get rid of manual dynamic linkage and leave it on ICD. It can hurt performance in some cases however.

Presumably, it will generate almost the same binary without extra overhead on hot path (e.g. extra indirection). But the patch makes the code uglier.

### CODE REVIEW

**Comment by stenzek** [comment]
Rather than duplicating the function pointers, it's probably simpler to just skip instance creation when the emulation is active. It's not possible to change the backend at runtime, anyway. This is likely an issue across all backends (e.g. D3D creates would create an additional device while rendering is occurring too).

**Comment by isilence** [comment]
I thought about it, but would it be okay with you? This bug appeared in the first place because of shared global state. Also, it would prevent from other potential features like multiple devices/instances usage.

**Comment by nathanraiderxxx** [comment]
Is the fix including also the android plugin? Cos for some device is still
buggy

Il Dom 16 Set 2018, 14:08 Pavel Begunkov <notifications@github.com> ha
scritto:

**Comment by JosJuice** [comment]
The issue that this pull request fixes does not exist on Android to begin with.

**Comment by isilence** [comment]
@nathanraiderxxx @JosJuice 
At least It covers all code paths related to vk library/function loading, so should work for android as well if that's the case.

**Comment by isilence** [comment]
@nathanraiderxxx 
There are also some other bugs, which would crush it on dynamic reconfiguring. Does it segfaulting just after you open a graphic configuration window/widget (whatever it is for android)? or after you changed some configs?

**Comment by nathanraiderxxx** [comment]
I'm following u guys cos I need to understand a but more of dolphin ..I
found some serious bugs in certain titles..and I know loads of users are
actually playing well...wich is weird...check out resident evil 4 for
example..RE4  (Gc and WII) BOTH SUFFER OF THE exact same issues in
VULKAN..In opengl..run well graphically..bit lag a lot in large areas..like
from 30 fpse goes randomly around the 20th

Il Dom 16 Set 2018, 14:31 Pavel Begunkov <notifications@github.com> ha
scritto:

**Comment by JosJuice** [comment]
@isilence The graphics configuration on Android doesn't call any native code, it just reads and writes INI files.

**Comment by nathanraiderxxx** [comment]
Trust me ...8 been seen loads of improvements guys.. this is possible..I
only wish I could help ..but so far I can only report bugs..by the way
sorry if I'm interrupting the conversation guys ..but I need to do some bug
report directly here..I'm FPSE developer..I rather be here and follow the
fixes u doing..hope u wont mind..but i see lot of potential i  dolphin
..and in 2 months i really notice many good improvements

Il Dom 16 Set 2018, 14:33 Pavel Begunkov <notifications@github.com> ha
scritto:

**Comment by JosJuice** [comment]
@nathanraiderxxx If you don't stop posting off-topic comments, you might get blocked from this repo.

**Comment by nathanraiderxxx** [comment]
Where do I ask help ? It suppose to be for help users...us..why this does
bother u so much?!I dknt understand..u make people hate dolphin if u speak
like this to them ....and isn't  the first time...wish I could see u in our
position and see ur reactions..think about it..and learn to be a bit more
soft speaking and available...wont hurt u.thank u

Il Dom 16 Set 2018, 14:40 JosJuice <notifications@github.com> ha scritto:

**Comment by JosJuice** [comment]
> Where do I ask help ? It suppose to be for help users...us..

This GitHub repo is explicitly not for helping users, it's for developers. The forums are the appropriate place for the kinds of comments you're posting.

**Comment by nathanraiderxxx** [comment]
I dont have bad intentions at all...I'm just trying to find any possible
way to help myself..and many other people affected with the same
problems..but if interacting with people offend u,then i give up..forums
are to help...but 90%of the time that people post new threads.. the answers
that some admin gives to us are really poor and non sens...I give up..enjoy
ur sunday evening and sorry for wasting ur times..all I wanted is to
collaborate with u..but understood.

Il Dom 16 Set 2018, 14:49 JosJuice <notifications@github.com> ha scritto:

**Comment by stenzek** [comment]
> I thought about it, but would it be okay with you? This bug appeared in the first place because of shared global state. Also, it would prevent us from implementing other potential features (e.g. multiple devices/instances).

Replacing global variables with members of a global struct is still global state. To be honest, I don't see any real need for multiple instances. But I'm happy to be convinced otherwise.

> Sure, it's easier, but won't solve the problem in a long run.

I'd rather fix the problem at its root, that is, the UI calling graphics functions and racing with the CPU/GPU thread. Rather than introducing additional code into API-specific backends.

**Comment by isilence** [comment]
> Replacing global variables with members of a global struct is still global state. To be honest, I don't see any real need for multiple instances. But I'm happy to be convinced otherwise.

It's not related to the bug, so it would have been sent by another patch.  

Multiple instances are useless as far as I know. But multi-device case for some compute/rendering offloading to let's say iGPU could be beneficial. Depends on the needs however.

> I'd rather fix the problem at its root, that is, the UI calling graphics functions and racing with the CPU/GPU thread. Rather than introducing additional code into API-specific backends.

Means making caps polling and rendering independent by purging shared state (altogether with g_Config).

But let me see what I can do with all of that.

**Comment by stenzek** [comment]
> Multiple instances are useless as far as I know. But multi-device case for some compute/rendering offloading to let's say iGPU could be beneficial. Depends on the needs however.

Unlike OpenGL where maintaining context function pointers is needed, does the Vulkan API not already support this? Most functions take a device/physical device parameter. It's my understanding that the loader would then dispatch to the correct ICD. So, it's basically duplicating the loader.

(note, yes, currently we do grab the device function directly, but if we did need cross-vendor/driver support, IMO it's better to just go through the officially supported path)

**Comment by isilence** [comment]
> Unlike OpenGL where maintaining context function pointers is needed, does the Vulkan API not already support this? Most functions take a device/physical device parameter. It's my understanding that the loader would then dispatch to the correct ICD. So, it's basically duplicating the loader.

vkGetDeviceProcAddr accepts device instance as an argument, means all returned pointers _may be_ device specific. So, that's true only if we use generic dispatch table, but not creating it by hand. 

tl;dr; 
It's so for performance reasons. AFAIK Linux ICD loader use two step dispatching: first any call goes to the loader, which then redirects it to an ICD. To get rid of extra indirect calls vkGetDeviceProcAddr may resolve function directly to a function from ICD.

https://vulkan.lunarg.com/doc/view/1.0.57.0/windows/LoaderAndLayerInterface.html#user-content-best-application-performance-setup

**Comment by isilence** [comment]
It may be off topic, but are there some performance tests/benchmarks for dolphin?

**Comment by isilence** [comment]
@stenzek, I didn't figured out how make it in a more elegant way, so there is your version. It uses coarse-grained synchronisation, but it's slow-path anyway.

NB: This makes it even more tightly coupled and error prone. IMHO before any major change It should be revised and refactored, otherwise we will see even more such bugs.

**Comment by stenzek** [comment]
I still think this issue is better fixed at the root of the problem - calling these functions from multiple threads. It doesn't make a lot of sense to me to make the Vulkan backend thread-safe here, but none of the others (I'm aware GL doesn't probe the context in InitBackendInfo, but it really should...).

**Comment by isilence** [comment]
Mmm, let me wrap it up. We have several concurrent threads (grapthics and GUI ones) populating shared info. There are several options I see:

1. To make InitBackendInfo reentrant, so letting it to be invoked independentely (*). That's what the first patchset have been doing -- GPU thread uses global functions pointers (packed into structure), but InitBackendInfo creates its own private copy of foo ptrs. IMHO, it's the most righteous way to solve the problem, just get rid of shared parts while we can.  
\* We also need either to synchronise shared ``` vulkan_module ```, make several module instances, or init it in advance (no lazy loading).

2. Synchronise InitBackendInfo (current patchset). InitBackendInfo called only on backend starting or some actions with config (e.g. first config open).

3. Making initialisation before running GPU thread. But assuming config can be opened at any moment, that's not a viable option.

4. other ideas?

If you have better ideas, please tell me (or implement it yourself). Now it crashes every second time I open the config window hovewer.

**Comment by isilence** [comment]
It would be great to have some rules about sync between threads...

```
bool LoadVulkanLibrary() { 
  // Not thread safe if a second thread calls the loader whilst the first is still in-progress.
```

\...because now we do exactly what the message is warning about.

**Comment by stenzek** [comment]
I'm not sure it needs this level of complexity. AFAIK, there are two paths which can call InitBackendInfo:

 - Opening the graphics window.
 - Booting the core (starting a game).

Those two cannot happen at the same time (the UI should be blocked until the CPU/GPU thread has started). Therefore, simply wrap the InitBackendInfo() call in the graphics dialog in a Core::IsRunning() check, and this issue shouldn't happen?

(of course, there is the issue of g_(Active)Config, which is not thread-safe, but that's unrelated to this specific crash)

**Comment by isilence** [comment]
Thank you for the explanation. I haven't looked at this code before.

Wrapping it in IsRunning also means that InitBackendInfo must be called only from GUI thread, otherwise we race again (because of possible Stop). There are some usages of InitBackendInfo, mainly in init and ShowConfig (I have no clue what it suppose to do). Maybe not racy, but too error-prone to leave it as is.

> but I can't think of any benefits which outweigh the additional indirection and complexity

* I wouldn't say it's complicated, but rather ugly. But it could be slightly refactored by embedding foo ptrs into Device, Render, etc classes and creating wrappers.  
* I don't have a clue what indirection do you mean. Render instead of calling ```(const_ptr)(...)```  would do ```(const_ptr + const_offset)(...)```, which will be compiled away.  
Moreover, 1-2 level indirect call into a library would hide any memory access overhead. But whatever...

**Comment by stenzek** [comment]
Dolphin has five video backends. If you were to do this properly, you would have to change every backend to pass a context around to most functions, or store a context pointer in every object (texture cache entry for example), which is a very significant change. And all this for a case which never occurs (requiring two contexts at the same time)?

I just can't see what the benefit is, when this is something which can be easily enforced as part of the contract between the frontend and core. There's plenty of other data structures which are touched by the UI in this manner, and don't cause any issues.
